sw/host/opentitanlib/src/console/spi.rs

```diff
@@ -0,0 +1,51 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+use anyhow::Result;
+use std::cell::Cell;
+use std::time::Duration;
+
+use crate::io::console::ConsoleDevice;
+use crate::io::spi::{Target, Transfer};
+
+pub struct SpiConsoleDevice<'a> {
+    pub spi: &'a dyn Target,
+    pub console_next_frame_number: Cell<u32>,
+}
+
+impl<'a> SpiConsoleDevice<'a> {
+    const SPI_FRAME_HEADER_SIZE: usize = 8;
+    const SPI_MAX_DATA_LENGTH: usize = 2032;
+}
+
+impl<'a> ConsoleDevice for SpiConsoleDevice<'a> {
+    fn console_read(&self, buf: &mut [u8], _timeout: Duration) -> Result<usize> {
+        // Read the SPI console frame header.
+        let mut header = vec![0u8; SpiConsoleDevice::SPI_FRAME_HEADER_SIZE];
+        self.spi
+            .run_transaction(&mut [Transfer::Write(&[0xff; 4]), Transfer::Read(&mut header)])?;
+        let frame_number: u32 = u32::from_le_bytes(*&header[0..4].try_into().unwrap());
+        let data_len_bytes: usize = u32::from_le_bytes(*&header[4..8].try_into().unwrap()) as usize;
+        if frame_number != self.console_next_frame_number.get()
+            || data_len_bytes > SpiConsoleDevice::SPI_MAX_DATA_LENGTH
+        {
+            // This frame is junk, so we do not read the data.
+            return Ok(0);
+        }
+        self.console_next_frame_number.set(frame_number + 1);
+
+        // Read the SPI console frame data.
+        let data_len_bytes_w_pad = data_len_bytes + (4 - (data_len_bytes % 4));
```

Comment:
Do you know the and-mask trick for rounding up to powers of two?
```
let n = <some power of 2 minus 1>;
let rounded_value = (value + n) & !n;  // `~n` in C, `!n` in rust.
```
In your case, it would be:
```
let data_len_bytes_w_pad = (data_len_bytes + 3) & !3;
```

Reply:
We are not rounding to a power of 2 here. We are rounding up to word size.

Reply:
Chris means that if you want to round up to the nearest multiple of a constant, and that constant happens to be a power of two, then you can use this trick.  In this case 4 is a power of two, so the trick works.

Reply:
I want to add that it took me a little while to figure out what you were trying to do with the modulus calculation, but since I know the same trick, I would instantly recognize what Chris' code is trying to do.  (This is a poor argument, since another engineer might be used to your way of expressing the rounding, and spend time figuring out what Chris' code is doing.)

Reply:
Hmm thats strange, because when I use Chris' code I get the wrong value.

Reply:
There might be something else going on, one moment.

Reply:
Ahh, my bad, I had to change the firmware side because I was always sending pad bytes (1--4, inclusive), I wasn't purely rounding up, I forgot, sorry. Changed to rounding up method. Thanks for the pro tip!

Reply:
Should be fixed now.